src/Chainweb/Mempool/Consensus.hs

```diff
@@ -0,0 +1,133 @@
+( chainwebTxsFromPWO
+, processFork
+, ReintroducedTxs (..)
+) where
+
+------------------------------------------------------------------------------
+import Streaming.Prelude (Of)
+import qualified Streaming.Prelude as S hiding (toList)
+
+import Control.DeepSeq
+import Control.Exception
+import Control.Monad
+import Control.Monad.Catch
+
+import Data.Aeson
+import Data.Either
+import Data.Foldable (toList)
+
+import Data.Set (Set)
+import qualified Data.Set as S
+import qualified Data.Text as T
+import Data.Vector (Vector)
+import qualified Data.Vector as V
+
+import GHC.Generics
+
+import System.LogLevel
+------------------------------------------------------------------------------
+import Chainweb.BlockHeader
+import Chainweb.BlockHeaderDB
+import Chainweb.Payload
+import Chainweb.Transaction
+import Chainweb.TreeDB
+import Chainweb.Utils
+
+import Data.LogMessage (JsonLog(..), LogFunction)
+------------------------------------------------------------------------------
+processFork
+    :: Ord x
+    => LogFunction
+    -> BlockHeaderDb
+    -> BlockHeader
+    -> Maybe BlockHeader
+    -> (BlockHeader -> IO (S.Set x))
+    -> IO (V.Vector x)
+processFork _ _ _ Nothing _ = return V.empty
+processFork logFun db newHeader (Just lastHeader) payloadLookup = do
+
+    let s = branchDiff db lastHeader newHeader
+    (oldBlocks, newBlocks) <- collectForkBlocks s
+    case V.length newBlocks - V.length oldBlocks of
+        n | n == 1    -> return V.empty -- no fork, no trans to reintroduce
+        n | n > 1     -> throwM $ MempoolConsensusException ("processFork -- height of new block is"
+                                      ++ "more than one greater than the previous new block request")
+          | otherwise -> do -- fork occurred, get the transactions to reintroduce
+              oldTrans <- foldM f mempty oldBlocks
+              newTrans <- foldM f mempty newBlocks
+              -- before re-introducing the transactions from the losing fork (aka oldBlocks), filter
+              -- out any transactions that have been included in the winning fork (aka newBlocks)
+
+              -- return $ V.fromList $ S.toList $ oldTrans `S.difference` newTrans
+              let results = V.fromList $ S.toList $ oldTrans `S.difference` newTrans
+
+              -- create data for the dashboard showing number or reintroduced transacitons:
+              let !reIntro = ReintroducedTxs
+                    { oldForkHeader = (ObjectEncoded lastHeader)
+                    , newForkHeader = (ObjectEncoded newHeader)
+                    , numReintroduced = V.length results
+                    }
+              logg Info $! "transactions reintroduced" <> sshow (V.length results)
+              logFun @(JsonLog ReintroducedTxs) Info $ JsonLog reIntro
+              return results
+  where
+    f trans header = S.union trans <$> payloadLookup header
+
+    logg :: LogLevel -> T.Text -> IO ()
+    logg = logFun
+
+
+data ReintroducedTxs = ReintroducedTxs
```

Comment:
Maybe add a suffix to indicate to the reader that this is intended for log messages?